src/shared/codelens/csharpCodeLensProvider.ts

```diff
@@ -0,0 +1,305 @@
+                            class: tuplet.class,
+                            method: methodSymbol,
+                        }
+                    })
+                )
+
+                return accumulator
+            },
+            []
+        )
+}
+
+export async function findParentProjectFile(
+    sourceCodeUri: vscode.Uri,
+    findWorkspaceFiles: typeof vscode.workspace.findFiles = vscode.workspace.findFiles,
+): Promise<vscode.Uri | undefined> {
+    const workspaceProjectFiles: vscode.Uri[] = await findWorkspaceFiles(
+        '**/*.csproj'
+    )
+
+    // Use the project file "closest" in the parent chain to sourceCodeUri
+    // Assumption: only one .csproj file will exist in a given folder
+    const parentProjectFiles = workspaceProjectFiles
+        .filter(uri => {
+            const dirname = dirnameWithTrailingSlash(uri.fsPath)
+
+            return sourceCodeUri.fsPath.startsWith(dirname)
+        })
+        .sort()
+        .reverse()
+
+    return parentProjectFiles.length === 0 ? undefined : parentProjectFiles[0]
+}
+
+export function isPublicClassSymbol(
+    document: Pick<vscode.TextDocument, 'getText'>,
+    symbol: vscode.DocumentSymbol,
+): boolean {
+    if (symbol.kind === vscode.SymbolKind.Class) {
+        const methodText: string = document.getText(symbol.range)
+
+        // Find the position of the 'class' keyword
+        const classPosition = methodText.indexOf('class')
+
+        if (classPosition !== -1) {
+            const classDecorationPrefix = methodText.substr(0, classPosition)
+
+            return METHOD_SYMBOL_PUBLIC_REGEXP.test(classDecorationPrefix)
+        }
+    }
+
+    return false
+}
+
+export function isPublicMethodSymbol(
+    document: Pick<vscode.TextDocument, 'getText'>,
+    symbol: vscode.DocumentSymbol,
+): boolean {
+    if (symbol.kind === vscode.SymbolKind.Method) {
+        const methodText: string = document.getText(symbol.range)
+
+        const methodName = getMethodNameFromSymbol(symbol)
+
+        if (!methodName) {
+            return false
+        }
+
+        // Look for the method name
+        const methodNameRegex = new RegExp(`\\b${methodName}\\b`)
+
+        // public async foo() -> "public async "
+        const signaturePosition = methodText.search(methodNameRegex)
+
+        if (signaturePosition === -1) {
+            const logger = getLogger()
+
+            const err = new Error(
+                `Unable to find function signature: ${methodName}`
+            )
+
+            logger.error(err)
+            throw err
+        }
+
+        const signatureBeforeMethodName = methodText.substr(0, signaturePosition)
+
+        return METHOD_SYMBOL_PUBLIC_REGEXP.test(signatureBeforeMethodName)
+    }
+
+    return false
+}
+
+export function getMethodNameFromSymbol(symbol: vscode.DocumentSymbol): string | undefined {
+    if (symbol.kind === vscode.SymbolKind.Method) {
+        const matches = symbol.name.match(METHOD_SYMBOL_NAME_REGEXP)
```

Comment:
In what case will a `SymbolKind.Method` symbol not match this regex? Just if the user hasn't typed it yet?